### ISSUE TITLE
PHP cgi autoflush not work

### ISSUE THREAD

**Issue by DaxMX** [civetweb/civetweb#1046]
The server only process the first autoflush from php scripts (not tested any other cgi scripts), in example, this simple test will fail:

```php
<?php
set_time_limit(20);
header('Content-Type: text/plain; charset=utf-8');
echo "CivetWeb Auto Flush Test\n";
for($i = 0; $i <10; $i++) {
	echo "Auto flush $i\n";
	@ob_flush();@flush();
	sleep(1);
}
echo "\nCurrently this test fails, only the first FLUSH is processed by the server.\n";
?>
```
The same result usin PHP 7.4.6, 8.0 and 8.1 with civetweb 1.15 (currently latest version).

Is there a setting to fix this?

**Comment by bel2125** [comment]
I tried to reproduce this, but I need some more information.

What server operating system are you using: Linux, Windows, ...?
What is your expected behavior. In my test environment (Server on Linux, Browser Edge in Windows), this page takes about 10 seconds to load, then the response is shown all at once:
```
CivetWeb Auto Flush Test
Auto flush 0
Auto flush 1
Auto flush 2
Auto flush 3
Auto flush 4
Auto flush 5
Auto flush 6
Auto flush 7
Auto flush 8
Auto flush 9

Currently this test fails, only the first FLUSH is processed by the server.
```
There is no intermediate result shown ... the page is not built line by line - neither the first line nor any other line. This is also true if I test with Google Chrome instead of Microsoft Edge.
 
What is your observation in your test environment, what exactly is your test environment, and what is your expectation?

**Comment by bel2125** [comment]
According to the comments in
https://www.php.net/manual/en/function.flush.php
https://www.php.net/manual/en/function.ob-flush.php

_As of August 2012, all browsers seem to show an all-or-nothing approach to buffering. In other words, while php is operating, no content can be shown._

Testing with curl, the first block is sent and then the rest. Following the discussions on PHP, flush is fundamentally broken there, and even if it would work there, there are several other components buffering as well. If you really need to send partial updates from PHP, there would be a small modification that allows for this, and some settings - all of them come at the cost of performance: set tcp_nodelay to 0, set allow_sendfile_call to no, and there is one compiler define to set the buffer size to one byte. But this will not help anything if the browser buffers everything until the page is complete.

**Comment by DaxMX** [comment]
Thank you for the reply...
I'm using windows 10 x64 for server and the result in any browser is the same (Tested locally from Android, iPad, Windows and Linux).

Expected result:
`Each echo to be displayed in the browser after 1 second`
Actual result:
`Fisrt echo displayed after one second. The rest of the result is displayed after php execution ends (9 seconds after).`

If I change civetweb for nginx, the test work as expected by set this in nginx.conf:
`fastcgi_buffering off;`
Or by adding an extra header in the php test script:
`header('X-Accel-Buffering: no');`

I'm looking for a little server to include with my multi platform web application and find civetweb as a nice candidate, but autoflush not work, is not php, is the way the server handles responses from fast-cgi application, I think an option like in nginx to support autoflush would be nice even to change a global define in the source and recompile the binary at the cost of performance.

By the way, I tried to set 'allow_sendfile_call no' in civetweb.conf and civetweb ignores it saying is an invalid option.

Anyway, civetweb is very very nice, congrats.

**Comment by bel2125** [comment]
"allow_sendfile_call" is a Linux only option - you don't need it for Windows (Windows does not have a sendfile call).

I exported a new option "cgi_buffering": https://github.com/civetweb/civetweb/blob/master/docs/UserManual.md#cgi_buffering-yes
You can set this to "no" - then the server will behave like you described.

In the latest release, this option was not exported but required to modify a define.
In the current head version I exported this option.

I still cannot make any guarantees that there is no intermediate buffer in some proxy or the browser. But this is not different for nginx - so if it worked in your test with nginx test environment then it will also work in the same environment with civetweb.

Since there is no released version with a pre-built binary yet:
Can you build your own version from using Visual Studio or do you want a beta version for your own tests?
A release date is not fixed yet - maybe January or February 2022.

**Comment by DaxMX** [comment]
Thank you for the new configuration option to allow unbuffered cgi responses.

I tested and now it works by set `cgi_buffering no` but has a serious performance impact, not at server or cgi execution level but at web browser level (using a similar test script to render some images), this is because civetweb sends the data to the web browser BYTE to BYTE and not as received from the cgi program.

By the way, on Windows works without set `tcp_nodelay 0` or by setting it to any value in civetweb.conf.

**Comment by bel2125** [comment]
Yes, it has a serious performance impact, as also documented in the [UserManual.md](https://github.com/civetweb/civetweb/blob/master/docs/UserManual.md#cgi_buffering-yes).

The implementation of reading byte by byte is the only one that is operating system independent. According to some web pages, it is even different between different members of the UNIX-Family, *completely* different for Windows, I don't know for MacOS, BSD and the various embedded operating systems supported by CivetWeb.

For Linux servers there might be a solution with ioctl/FIONREAD or select - but this will definitely not work for Windows servers (for Windows there are some suggestions to use PeekNamedPipe, while other comments state this will not work). Even if we add some system specific optimization for a selected system, this byte by byte reading will always the backup plan for systems that do not support any other solution. An additional problem: even if I add an optimization, I cannot test if it works on every operating system or if this optimization completely breaks the feature for a particular system. I can test for Windows10 and some major Linux distributions (like Ubuntu/Debian) but not for every possible operating system.

I am not sure if using a HEAVILY operating system dependent optimization is justified for this. Even if this feature works from the webserver, a proxy/cache or the browser may still buffer, so the whole chain still does not work and every effort invested into the server would be a waste or time. An alternative would be to use websockets for asynchronous contents. All intermediate components handle this protocol correctly. Even if there is no other buffer in the chain (it seems this means the client can not be a web browser - since they usually buffer the full page), probably only a small portion of the php scripts send partial contents like the test script above. Then it would still be possible to turn of buffering only for the few scripts that do send partial contents and keep buffering for the many other scripts (can be done by using a different file extension and the cgi2_buffering configuration).

Or maybe I am missing a simple, preferably operating system independent method how this can be done - although I have been looking for several hours. If there is such a method, please tell me.
